### ISSUE TITLE
Mint Unlock Discount Tokens

### ISSUE THREAD

**Issue by NickCuso** [unlock-protocol/unlock#4282]
We have an upgradable ERC-20 implementation for the Unlock Discount Token.  Once we have #3711 implemented we can begin minting UDT with each Key purchase.

Just need to lock down the formula we want to start with.

After this is complete we can talk about applying a discount based on the user's holdings.

**Comment by julien51** [comment]
It's now time to work on this!
[This wiki page](https://github.com/unlock-protocol/unlock/wiki/The-Unlock-Discount-Tokens) provides the description of the Unlock Discount Token.

The minting process (creation of new tokens) happens through a single mechanism: referrals.
Every time a new key is purchased if there is a referrer the Unlock smart contract's `recordKeyPurchase` method will be invoked by the lock.

This method already implements the logic which consists of increasing the `grossNetworkProduct`. We now need to add the logic of granting tokens to the `referrer`.

There are 2 rules which help determine the number of tokens minted for the referrer.

1. Gas limit
In order to avoid Sybil attacks (lock owners referring themselves), the value of tokens earned is AT MOST equal to the value of gas spent.
In other words, if someone spent 0.001 Eth worth of gas, they should earn an amount of UDT which is at most 0.001 Eth.

The value of UDT should be retrieved from a Uniswap pool. Unlock Inc will provide liquidity for such a pool, with a very low price, assuming that the tokens actually have little value until the "discount claiming" mechanism is implemented.

2. Token inflation
Additionally, in order to guarantee "scarcity" of the Unlock Discount Tokens, the number of tokens minted/granted is such that the supply of tokens increases AT MOST by half of the increase in GDP resulting from the purchase. In other words, if the `grossNetworkProduct` increases by 1% when a referral is made, the number of tokens minted should grow the supply by at most 0.5%.

**Comment by julien51** [comment]
Here are things we need to clarify/set before we set this in montion:

[ ] What is the address of the ERC20 UDT contract?
[ ] We should pre-mine at least 1 token (in order to perform the calculation when it comes to supply.
[ ] We should set a Uniswap pool for UDT/Eth.

**Comment by julien51** [comment]
Here are the tasks which I am identifying for this to be fully implemented:

- [ ] Retrieve the UDT price in Eth from the Uniswap pool
- [ ] Compute Eth spent for gas when purchasing a key (gas amount could be hardcoded while gas price needs to be actual value)
- [ ] Determine max amount of tokens to be granted based on gas spent
- [ ] Compute growth of `grossNetworkProduct` in %
- [ ] Compute max growth of UDT supply based on `grossNetworkProduct` growth.
- [ ] Use minimum of the 2 calculation to determine the actual number of tokens to be granted
- [ ] Mint and grant ERC20 to referrer.

**Comment by NickCuso** [comment]
Some comments / questions (no particular order) @julien51 

 - gas limit: we won't be able to know exactly how much gas was consumed. We could hard code an estimate of the gas required and lookup the gasPrice used for the transaction. The largest varience would probably be if one of the hook callbacks are used which will use an unknown amount.

Solidity does allow us to check `gas` which is the amount still available for execution. We could check that at the beginning of the purchase call and again at the end, as low in the execution flow as possible before we mint tokens. If we update PublicLock a bit we could ensure the ERC-20 transfer and hook callback complete first - getting us close to the actual value used. For this we would need a small change to the Unlock.sol function as well (in order to pass the original `gas` value).

 - Presumably we will use UniswapV2 for this. We may want to take on the task to upgrade our GDP calculation to use UniswapV2 first.

 - The UDT contract was deployed, the address is in the .openzeppelin directory. It was not verified though. If there's any challenge with that we could just redeploy since it's not yet been used.  Additionally we may want to add EIP-2612 support for the token (that said we are using an upgradable proxy so this could be added later on).

 - The pre-mine of just 1 token may cause a lot of volatility at first. We may want to kick off the Uniswap pool with a bit more.

 - Will we mint the dev reward as well (mentioned on the wiki but not in the checklist above)

 - The token freeze applies to transfer as well? It seems that may be best, but would require we put the logic into the token contract instead of in Unlock.sol.  Additionally maybe we use timespan instead of block count. I wonder if the move to ETH2 will have an impact on the average time between blocks. Timespan may be closer to the feature intent.

 - Do not mint for renewals: we don't have this information in Unlock.sol today. We can add this but would require some change to PublicLock.

**Comment by julien51** [comment]
Thanks @NickCuso !

1. Gas limit: ok to hardcode. It seems easier and should cover our base case. I think this should be "updateable" by the Unlock owner.

2. Good point about upgrading to Uniswap v2! Let's do this too.

3. Let's verify the contract unless you think we need something else? Since it's upgradable, I believe we can still make changes to it later?

4. Regarding the pre-mine, I guess we can probably increase it then. I believe it should not affect the protocol since eventually the number of coins minted is based on the growth of the network. 
I did a bunch of simulation work last night and I the pre-mined amount itself is not significant, as the supply will grow with the GDP itself. Let's pre-mine a total of 1,000,000 tokens!

5. Yes, we should absolutely mine the dev reward. 20% of each minted token goes to the Unlock owner address.

6. The token freeze should apply, but I think we should only implement it when we actually have the discount mechanism? For now, we can let transfers happen freely as they do not have any impact on the actual value of the tokens.

7. Couldn't we check if there was previously a key owned by that address on the lock?

**Comment by NickCuso** [comment]
Sounds good. RE 7) not without changing public lock. The current PublicLock implementation issues the NFT before calling Unlock - so I don't see a way to know if it's a renewal or not.

**Comment by julien51** [comment]
Dang. Well let's grant even for renewals then.

**Comment by NickCuso** [comment]
Okay. To be clear, we could fix this but it wouldn't be backwards compatible with older PublicLock versions. e.g. maybe we grant for all purchases for old versions and only for new purchases when PublicLock is >= v8

**Comment by julien51** [comment]
Let's play it simple for now and not change PublicLock.

**Comment by NickCuso** [comment]
After upgrading we will need to:
 - [ ] Mint UDT to fund the Uniswap pool
 - [ ] Add Unlock.sol as a UDT minter (and maybe remove the original account)
 - [ ] Approve the uniswapRouter & create a UDT - WETH UniswapV2 pool
 - [ ] Make a trade on Uniswap so that the oracle can determine the exchange rate
 - [ ] Call configUnlock on Unlock to set the UDT address and the estimated gas for a purchase
 - [ ] Call setOracle on Unlock to add support for UDT

**Comment by julien51** [comment]
All done :)
